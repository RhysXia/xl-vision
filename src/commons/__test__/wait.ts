import { act } from 'react-dom/test-utils'

export default (time: number) => {
  const wait = new Promise<void>(resolve => {
    // console.info(`wait time: ${time}`)
    setTimeout(() => {
      resolve()
    }, time)
  })

  return act(() => wait)
}
