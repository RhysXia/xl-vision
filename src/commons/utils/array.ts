export function oneOf<T>(array: T[], item: T) {
  for (const _item of array) {
    if (_item === item) {
      return true
    }
  }
  return false
}
