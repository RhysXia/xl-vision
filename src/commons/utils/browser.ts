import { isServer } from './env'

// 判断是否支持指定的css属性
// eslint-disable-next-line import/prefer-default-export
export const isStyleSupport = (styleProperty: string) => {
  if (isServer) {
    return true
  }
  const styles = document.body.style
  let result = false
  if (styleProperty in styles) {
    result = true
  } else {
    const vendors = ['ms', 'Webkit', 'Moz', 'O']
    for (let i = 0; i < vendors.length; i++) {
      const vendor = vendors[i]
      const prop = vendor + styleProperty.replace(/^[a-z]/, (val) => val.toUpperCase())
      if (prop in styles) {
        result = true
        break
      }
    }
  }
  return result
}
