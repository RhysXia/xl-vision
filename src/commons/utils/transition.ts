import { isClient } from './env'

let transitionProp = 'transition'
let transitionEndEvent = 'transitionend'
let animationProp = 'animation'
let animationEndEvent = 'animationend'

if (isClient) {
  if (window.ontransitionend === undefined &&
    // @ts-ignore
    window.onwebkittransitionend !== undefined
  ) {
    transitionProp = 'WebkitTransition'
    transitionEndEvent = 'webkitTransitionEnd'
  }
  if (window.onanimationend === undefined &&
    // @ts-ignore
    window.onwebkitanimationend !== undefined
  ) {
    animationProp = 'WebkitAnimation'
    animationEndEvent = 'webkitAnimationEnd'
  }
}

export const onTransitionEnd = (el: HTMLElement, done: () => void) => {
  const styles = getComputedStyle(el)
  // @ts-ignore
  const transitionDelays: string[] = (styles[`${transitionProp}Delay`] || '').split(', ')
  // @ts-ignore
  const transitionDurations: string[] = (styles[`${transitionProp}Duration`] || '').split(', ')
  // @ts-ignore
  const animationDelays: string[] = (styles[`${animationProp}Delay`] || '').split(', ')
  // @ts-ignore
  const animationDurations: string[] = (styles[`${animationProp}Duration`] || '').split(', ')
  const transitionTimeout: number = getTimeout(transitionDelays, transitionDurations)
  const animationTimeout: number = getTimeout(animationDelays, animationDurations)
  let event = transitionEndEvent
  let timeout = transitionTimeout
  let eventCount = transitionDurations.length

  if (timeout < animationTimeout) {
    event = animationEndEvent
    timeout = animationTimeout
    eventCount = animationDurations.length
  }

  if (timeout <= 0) {
    return done()
  }

  let count = 0

  const end = () => {
    el.removeEventListener(event, onEnd)
    done()
  }

  const onEnd = (e: Event) => {
    if (e.target !== el) {
      return
    }
    if (++count >= eventCount) {
      end()
    }
  }

  setTimeout(() => {
    if (count < eventCount) {
      end()
    }
  }, timeout + 1)

  el.addEventListener(event, onEnd)
}

const getTimeout = (delays: string[], durations: string[]) => {
  while (delays.length < durations.length) {
    delays = delays.concat(delays)
  }

  return Math.max.apply(null, durations.map((d, i) => {
    return toMs(d) + toMs(delays[i])
  }))
}

const toMs = (s: string) => {
  return Number(s.slice(0, -1)) * 1000
}

const raf = isClient ?
  window.requestAnimationFrame ?
    window.requestAnimationFrame.bind(window)
    : setTimeout
  : (fn: Function) => fn()

export const nextFrame = (fn: Function) => {
  raf(() => {
    raf(fn)
  })
}
