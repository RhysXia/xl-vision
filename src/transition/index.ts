import CSSTransition from './css-transition'
import Transition from './transition'
import TransitionGroup from './transition-group'

export { TransitionProps } from './transition'
export { CSSTransitionProps } from './css-transition'
export { TransitionGroupProps } from './transition-group'

export {
  CSSTransition,
  Transition,
  TransitionGroup
}

export default Transition
