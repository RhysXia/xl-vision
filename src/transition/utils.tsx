import { ProxyElement } from '.'

export const proxyElement = (el: HTMLElement, isWriteable: () => boolean): ProxyElement => {
  const setHandler = (target: any, prop: PropertyKey, value: any, _receiver?: any) => {
    if (isWriteable()) {
      target[prop] = value
    }
    return true
  }

  const getHandler = (target: any, prop: PropertyKey, _receiver?: any) => {
    return target[prop]
  }

  const style = new Proxy(el.style, {
    get: getHandler,
    set: setHandler
  })
  const classList = new Proxy(el.classList,{
    get: getHandler,
    set: setHandler
  })

  return new Proxy(el, {
    get: (target, prop, _receiver) => {
      if (prop === 'target') {
        return target
      }
      if (prop === 'style') {
        return style
      }
      if (prop === 'classList') {
        return classList
      }
      return getHandler(target, prop, target)
    },
    set: setHandler
  })
}
