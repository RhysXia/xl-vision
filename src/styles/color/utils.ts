import { warning } from '../../commons/utils/logger'

export type RGBColor = {
  r: number
  g: number
  b: number
  a: number
  type: 'rgb'
}

export type HSLColor = {
  hexToColor: number
  s: number
  l: number
  a: number
  type: 'hsl'
}

export type Color = RGBColor | HSLColor

const hex3 = /^#?([\dA-Fa-f])([\dA-Fa-f])([\dA-Fa-f])$/
const hex4 = /^#?([\dA-Fa-f])([\dA-Fa-f])([\dA-Fa-f])([\dA-Fa-f])$/
const hex6 = /^#?([\dA-Fa-f]{2})([\dA-Fa-f]{2})([\dA-Fa-f]{2})$/
const hex8 = /^#?([\dA-Fa-f]{2})([\dA-Fa-f]{2})([\dA-Fa-f]{2})([\dA-Fa-f]{2})$/

// <http://www.w3.org/TR/css3-values/#integers>
const CSS_INTEGER = '[-\\+]?\\d+%?'

// <http://www.w3.org/TR/css3-values/#number-value>
const CSS_NUMBER = '[-\\+]?\\d*\\.\\d+%?'

// Allow positive/negative integer/number.  Don't capture the either/or, just the entire outcome.
const CSS_UNIT = `(?:${CSS_NUMBER})|(?:${CSS_INTEGER})`

// Actual matching.
// Parentheses and commas are optional, but not required.
// Whitespace can take the place of commas or opening paren
const PERMISSIVE_MATCH3 = `[\\s|\\(]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})\\s*\\)?`
const PERMISSIVE_MATCH4 = `[\\s|\\(]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})\\s*\\)?`

const hsl = new RegExp(`hsl${PERMISSIVE_MATCH3}`)
const hsla = new RegExp(`hsla${PERMISSIVE_MATCH4}`)

/**
 * Converts a color from CSS hex format to CSS rgb format.
 *
 * @param {string} color - Hex color, i.e. #nnn or #nnnnnn
 * @returns {string} A CSS rgb color string
 */
export const hexToColor = (color: string) => {
  let match = hex8.exec(color)
  if (match) {
    return {
      r: parseInt(match[1], 16),
      g: parseInt(match[2], 16),
      b: parseInt(match[3], 16),
      a: parseInt(match[4], 16)
    }
  }

  match = hex6.exec(color)
  if (match) {
    return {
      r: parseInt(match[1], 16),
      g: parseInt(match[2], 16),
      b: parseInt(match[3], 16)
    }
  }

  match = hex4.exec(color)
  if (match) {
    return {
      r: parseInt(`${match[1]}${match[1]}`, 16),
      g: parseInt(`${match[1]}${match[2]}`, 16),
      b: parseInt(`${match[1]}${match[3]}`, 16),
      a: parseInt(`${match[1]}${match[4]}`, 16)
    }
  }

  match = hex3.exec(color)
  if (match) {
    return {
      r: parseInt(`${match[1]}${match[1]}`, 16),
      g: parseInt(`${match[1]}${match[2]}`, 16),
      b: parseInt(`${match[1]}${match[3]}`, 16)
    }
  }

  warning(true, `The color '%s' is illegal hex color.`, color)
}

const hueToRgb = (p: number, q: number, t: number) => {
  const _p = p
  const _q = q
  let _t = t
  if (_t < 0) _t += 1
  if (_t > 1) _t -= 1
  if (_t < 1 / 6) return _p + (_q - _p) * 6 * _t
  if (_t < 1 / 2) return _q
  if (_t < 2 / 3) return _p + (_q - _p) * (2 / 3 - _t) * 6
  return _p
}

export const hslToRgb = (h: string, s: string, l: string) => {
  const _h = parseFloat(h) / 360
  const _s = parseFloat(s) / 100
  const _l = parseFloat(l) / 100

  if (s === 0) {
    r = g = b = l // achromatic
  } else {
    var q = l < 0.5 ? l * (1 + s) : l + s - l * s
    var p = 2 * l - q
    r = hue2rgb(p, q, h + 1 / 3)
    g = hue2rgb(p, q, h)
    b = hue2rgb(p, q, h - 1 / 3)
  }

  return { r: r * 255, g: g * 255, b: b * 255 }
}

export const hslToColor = (color: string) => {
  let match = hsl.exec(color)
  if (match) {
    return { h: match[1], s: match[2], l: match[3] }
  }

  match = hsla.exec(color)
  if (match) {
    return { h: match[1], s: match[2], l: match[3], a: match[4] }
  }
}
