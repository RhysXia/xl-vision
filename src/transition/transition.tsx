import PropTypes from 'prop-types'
import * as React from 'react'
import { namePrefix } from '../commons/config'
import { proxyElement } from './utils'

export const STATE_INIT = 'init'
export const STATE_APPEARING = 'appearing'
export const STATE_APPEARED = 'appeared'
export const STATE_ENTERING = 'entering'
export const STATE_ENTERED = 'entered'
export const STATE_LEAVING = 'leaving'
export const STATE_LEAVED = 'leaved'

export type State =
  | typeof STATE_INIT
  | typeof STATE_APPEARING
  | typeof STATE_APPEARED
  | typeof STATE_ENTERING
  | typeof STATE_ENTERED
  | typeof STATE_LEAVING
  | typeof STATE_LEAVED

export interface ProxyElement extends HTMLElement {
  target: HTMLElement
}

export interface TransitionProps {
  afterAppear?: (el: HTMLElement) => void
  afterEnter?: (el: HTMLElement) => void
  afterLeave?: (el: HTMLElement) => void
  appear?: (el: ProxyElement, done: (() => void)) => void
  appearCancelled?: (el: HTMLElement) => void
  beforeAppear?: (el: HTMLElement) => void
  beforeEnter?: (el: HTMLElement) => void
  beforeLeave?: (el: HTMLElement) => void
  children: React.ReactElement
  enter?: (el: ProxyElement, done: (() => void)) => void
  enterCancelled?: (el: HTMLElement) => void
  in: boolean
  isAppear?: boolean
  leave?: (el: ProxyElement, done: (() => void)) => void
  leaveCancelled?: (el: HTMLElement) => void
  unmountOnLeave?: boolean
}

const displayName = `${namePrefix}-transition`

const Transition: React.FunctionComponent<TransitionProps> = props => {
  const {
    in: inProp,
    // 初次挂载时，如果是进入状态，是否触发appear动画
    isAppear,
    afterEnter,
    afterLeave,
    beforeEnter,
    beforeLeave,
    children,
    enter,
    enterCancelled,
    leave,
    leaveCancelled,
    unmountOnLeave
  } = props

  let { beforeAppear, appear, appearCancelled, afterAppear } = props

  // 如果开启appear,默认使用enter的生命周期方法
  beforeAppear = beforeAppear || beforeEnter
  appear = appear || enter
  afterAppear = afterAppear || afterEnter
  appearCancelled = appearCancelled || enterCancelled

  const [state, setState] = React.useState<State>(STATE_INIT)

  // 标记当前组件是否被卸载
  let isMounted = true

  React.useEffect(() => {
    // 卸载时设置标记isMounted = false
    return () => {
      isMounted = false
    }
  }, [])

  const childrenRel = React.useRef<HTMLElement>()

  const display = React.useMemo(() => {
    // 还未初始化完成
    if (state === STATE_INIT) {
      return false
    }
    if (unmountOnLeave && state === STATE_LEAVED) {
      return false
    }
    // 默认展示
    return true
  }, [state, unmountOnLeave])

  React.useEffect(() => {
    if (inProp) {
      // 此时说明leave动画还没有完成，需要触发leaveCancelled
      if (state === STATE_LEAVING) {
        leaveCancelled && leaveCancelled(childrenRel.current!)
      }
      // 还没有初始化
      if (state === STATE_INIT) {
        // 判断是否需要触发isAppear
        if (isAppear) {
          setState(STATE_APPEARING)
        } else {
          setState(STATE_ENTERED)
        }
      } else {
        setState(STATE_ENTERING)
      }
    } else {
      // 此时说明appear动画还没有完成，需要触发appearCancelled
      if (state === STATE_APPEARING) {
        appearCancelled && appearCancelled(childrenRel.current!)
      // 此时说明enter动画还没有完成，需要触发enterCancelled
      } else if (state === STATE_ENTERING) {
        enterCancelled && enterCancelled(childrenRel.current!)
      }
      // STATE_INIT只有在初始化才存在，所以要排除
      if (state === STATE_INIT) {
        setState(STATE_LEAVED)
      } else {
        setState(STATE_LEAVING)
      }
    }
  }, [inProp])

  React.useEffect(() => {
    if (state === STATE_APPEARING) {
      beforeAppear && beforeAppear(childrenRel.current!)
      onTransitionEnd(() => {
        afterAppear && afterAppear(childrenRel.current!)
        setState(STATE_APPEARED)
      }, appear)
      // 当前是离开或者正在离开状态，下一个状态为STATE_ENTERING
    } else if (state === STATE_ENTERING) {
      beforeEnter && beforeEnter(childrenRel.current!)
      onTransitionEnd(() => {
        afterEnter && afterEnter(childrenRel.current!)
        setState(STATE_ENTERED)
      }, enter)
    } else if (state === STATE_LEAVING) {
      beforeLeave && beforeLeave(childrenRel.current!)
      onTransitionEnd(() => {
        afterLeave && afterLeave(childrenRel.current!)
        setState(STATE_LEAVED)
      }, leave)
    }
  }, [state])

  const onTransitionEnd = React.useMemo(() => {
    let count = 0
    return (callback: () => void, action?: (el: ProxyElement, cb: (() => void)) => void) => {
      count++
      const match = count
      const isWriteable = () => match === count
      const ele = proxyElement(childrenRel.current!, isWriteable)
      const wrapCallback = () => {
        if (isWriteable() && isMounted) {
          callback()
        }
      }
      if (action) {
        action(ele, wrapCallback)
      } else {
        wrapCallback()
      }
    }
  }, [])

  const cloneChildren = React.useMemo(() => {
    return React.cloneElement(children, {
      ref: childrenRel
    })
  }, [children])

  return display ? cloneChildren : null
}

Transition.displayName = displayName

Transition.propTypes = {
  afterAppear: PropTypes.func,
  afterEnter: PropTypes.func,
  afterLeave: PropTypes.func,
  appear: PropTypes.func,
  appearCancelled: PropTypes.func,
  beforeAppear: PropTypes.func,
  beforeEnter: PropTypes.func,
  beforeLeave: PropTypes.func,
  children: PropTypes.element.isRequired,
  enter: PropTypes.func,
  enterCancelled: PropTypes.func,
  in: PropTypes.bool.isRequired,
  isAppear: PropTypes.bool,
  leave: PropTypes.func,
  leaveCancelled: PropTypes.func,
  unmountOnLeave: PropTypes.bool
}

export default Transition
