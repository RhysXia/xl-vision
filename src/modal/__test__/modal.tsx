import { mount } from 'enzyme'
import React from 'react'
import Modal from '..'
import { act } from 'react-dom/test-utils'
import * as TransitionUtils from '../../commons/utils/transition'
import wait from '../../../test/wait'

const nextFrameSpy = jest.spyOn(TransitionUtils, 'nextFrame')
nextFrameSpy.mockImplementation(fn => fn())

describe('Modal', () => {
  it('测试visible和onVisibleChange', () => {
    const fn = jest.fn()
    const wrapper = mount(
      <Modal visible={false} title={'title'} onVisibleChange={fn}>
        content
      </Modal>
    )
    expect(wrapper.render()).toMatchSnapshot()
    expect(fn.mock.calls.length).toBe(0)

    wrapper.setProps({ visible: true })
    wrapper.update()

    expect(fn.mock.calls[0][0]).toBe(true)

    expect(wrapper.render()).toMatchSnapshot()

    wrapper.setProps({ visible: false })
    wrapper.update()
    expect(fn.mock.calls[1][0]).toBe(false)
  })

  it('测试forceRender', () => {
    const wrapper = mount(
      <Modal visible={false} title={'title'} forceRender={true}>
        content
      </Modal>
    )
    expect(wrapper.render()).toMatchSnapshot()

    wrapper.setProps({ visible: true })

    wrapper.update()

    expect(wrapper.render()).toMatchSnapshot()
  })

  it('测试maskClosable', async () => {
    const wrapper = mount(
      <Modal visible={true} title={'title'} maskClosable={false}>
        content
      </Modal>
    )

    expect(wrapper.getDOMNode<HTMLDivElement>().style.display).toBe('')

    await act(() => {
      wrapper
        .find('.xl-modal__wrap')
        .at(0)
        .simulate('click')
      return wait(10)
    })

    expect(wrapper.getDOMNode<HTMLDivElement>().style.display).toBe('')

    wrapper.setProps({
      maskClosable: true
    })
    wrapper.update()

    expect(wrapper.getDOMNode<HTMLDivElement>().style.display).toBe('')

    await act(() => {
      wrapper
        .find('.xl-modal__wrap')
        .at(0)
        .simulate('click')
      return wait(10)
    })

    expect(wrapper.getDOMNode<HTMLDivElement>().style.display).toBe('none')
  })

  it('测试title', () => {
    const wrapper = mount(
      <Modal visible={true} title={'title'}>
        content
      </Modal>
    )

    expect(wrapper.render()).toMatchSnapshot()

    wrapper.setProps({
      title: <div>title</div>
    })

    wrapper.update()
    expect(wrapper.render()).toMatchSnapshot()
  })

  it('测试onOk和onCancel', async () => {
    const onOk = jest.fn()
    const onCancel = jest.fn()
    const wrapper = mount(
      <Modal visible={true} onOk={onOk} onCancel={onCancel}>
        content
      </Modal>
    )

    expect(onCancel.mock.calls.length).toBe(0)
    expect(onOk.mock.calls.length).toBe(0)

    // 点击mask触发onCancel
    await act(() => {
      wrapper
        .find('.xl-modal__wrap')
        .at(0)
        .simulate('click')
      return wait(10)
    })

    expect(onCancel.mock.calls.length).toBe(1)
    expect(onOk.mock.calls.length).toBe(0)
    onOk.mockClear()
    onCancel.mockClear()

    // 点击右上角close触发onCancel
    wrapper.setProps({
      visible: true
    })
    wrapper.update()
    expect(onCancel.mock.calls.length).toBe(0)
    expect(onOk.mock.calls.length).toBe(0)

    await act(() => {
      wrapper
        .find('.xl-modal__icon')
        .at(0)
        .simulate('click')
      return wait(10)
    })

    expect(onCancel.mock.calls.length).toBe(1)
    expect(onOk.mock.calls.length).toBe(0)
    onOk.mockClear()
    onCancel.mockClear()

    // 点击取消按钮触发onCancel
    wrapper.setProps({
      visible: true
    })
    wrapper.update()
    expect(onCancel.mock.calls.length).toBe(0)
    expect(onOk.mock.calls.length).toBe(0)

    await act(() => {
      wrapper
        .find('.xl-modal__footer button')
        .at(0)
        .simulate('click')
      return wait(10)
    })

    expect(onCancel.mock.calls.length).toBe(1)
    expect(onOk.mock.calls.length).toBe(0)
    onOk.mockClear()
    onCancel.mockClear()

    // 点击确认按钮触发onOk
    wrapper.setProps({
      visible: true
    })
    wrapper.update()
    expect(onCancel.mock.calls.length).toBe(0)
    expect(onOk.mock.calls.length).toBe(0)

    await act(() => {
      wrapper
        .find('.xl-modal__footer button')
        .at(1)
        .simulate('click')
      return wait(10)
    })

    expect(onCancel.mock.calls.length).toBe(0)
    expect(onOk.mock.calls.length).toBe(1)
  })

  it('测试closable', () => {
    const wrapper = mount(<Modal visible={true}>content</Modal>)

    expect(wrapper.find('.xl-modal__icon').length).toBe(1)

    wrapper.setProps({
      closable: false
    })
    wrapper.update()
    expect(wrapper.find('.xl-modal__icon').length).toBe(0)
  })

  it('测试destroyOnClose和afterClose', () => {
    const afterClose = jest.fn()
    const wrapper = mount(
      <Modal visible={true} destroyOnClose={false} afterClose={afterClose}>
        content
      </Modal>
    )

    expect(wrapper.find('.xl-modal__wrap').length).toBe(1)
    expect(afterClose.mock.calls.length).toBe(0)

    wrapper.setProps({
      visible: false
    })

    wrapper.update()

    // destoryOnClose=false, 所以节点不会删除
    expect(wrapper.find('.xl-modal__wrap').length).toBe(1)
    expect(afterClose.mock.calls.length).toBe(1)
    afterClose.mockClear()

    wrapper.setProps({
      visible: true,
      destroyOnClose: true
    })
    wrapper.update()

    expect(wrapper.find('.xl-modal__wrap').length).toBe(1)
    expect(afterClose.mock.calls.length).toBe(0)

    wrapper.setProps({
      visible: false
    })
    wrapper.update()

    expect(wrapper.find('.xl-modal__wrap').length).toBe(0)
    expect(afterClose.mock.calls.length).toBe(1)
  })
})
