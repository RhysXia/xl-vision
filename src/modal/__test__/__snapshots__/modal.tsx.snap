// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal 测试forceRender 1`] = `
<div
  class="xl-modal__wrap"
  style="z-index: 0; display: none;"
>
  <div
    class="xl-modal"
    style="width: 520px;"
  >
    <button
      class="xl-modal__icon"
    >
      <span
        class="xl-modal__icon-x"
      >
        <i
          class="xl-icon"
        >
          <svg
            fill="currentColor"
            style="height: 1em; width: 1em;"
            viewBox="0 0 352 512"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
        </i>
      </span>
    </button>
    <div
      class="xl-modal__header"
    >
      <div
        class="xl-modal__title"
      >
        title
      </div>
    </div>
    <div
      class="xl-modal__body"
    >
      content
    </div>
    <div
      class="xl-modal__footer"
    >
      <button
        class="xl-button xl-button--default xl-button--size-default"
      >
        <span>
          取消
        </span>
      </button>
      <button
        class="xl-button xl-button--primary xl-button--size-default"
      >
        <span>
          确定
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Modal 测试forceRender 2`] = `
<div
  class="xl-modal__wrap"
  style="z-index: 1002;"
>
  <div
    class="xl-modal"
    style="width: 520px;"
  >
    <button
      class="xl-modal__icon"
    >
      <span
        class="xl-modal__icon-x"
      >
        <i
          class="xl-icon"
        >
          <svg
            fill="currentColor"
            style="height: 1em; width: 1em;"
            viewBox="0 0 352 512"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
        </i>
      </span>
    </button>
    <div
      class="xl-modal__header"
    >
      <div
        class="xl-modal__title"
      >
        title
      </div>
    </div>
    <div
      class="xl-modal__body"
    >
      content
    </div>
    <div
      class="xl-modal__footer"
    >
      <button
        class="xl-button xl-button--default xl-button--size-default"
      >
        <span>
          取消
        </span>
      </button>
      <button
        class="xl-button xl-button--primary xl-button--size-default"
      >
        <span>
          确定
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Modal 测试title 1`] = `
<div
  class="xl-modal__wrap"
  style="z-index: 1004;"
>
  <div
    class="xl-modal"
    style="width: 520px;"
  >
    <button
      class="xl-modal__icon"
    >
      <span
        class="xl-modal__icon-x"
      >
        <i
          class="xl-icon"
        >
          <svg
            fill="currentColor"
            style="height: 1em; width: 1em;"
            viewBox="0 0 352 512"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
        </i>
      </span>
    </button>
    <div
      class="xl-modal__header"
    >
      <div
        class="xl-modal__title"
      >
        title
      </div>
    </div>
    <div
      class="xl-modal__body"
    >
      content
    </div>
    <div
      class="xl-modal__footer"
    >
      <button
        class="xl-button xl-button--default xl-button--size-default"
      >
        <span>
          取消
        </span>
      </button>
      <button
        class="xl-button xl-button--primary xl-button--size-default"
      >
        <span>
          确定
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Modal 测试title 2`] = `
<div
  class="xl-modal__wrap"
  style="z-index: 1004;"
>
  <div
    class="xl-modal"
    style="width: 520px;"
  >
    <button
      class="xl-modal__icon"
    >
      <span
        class="xl-modal__icon-x"
      >
        <i
          class="xl-icon"
        >
          <svg
            fill="currentColor"
            style="height: 1em; width: 1em;"
            viewBox="0 0 352 512"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
        </i>
      </span>
    </button>
    <div
      class="xl-modal__header"
    >
      <div>
        title
      </div>
    </div>
    <div
      class="xl-modal__body"
    >
      content
    </div>
    <div
      class="xl-modal__footer"
    >
      <button
        class="xl-button xl-button--default xl-button--size-default"
      >
        <span>
          取消
        </span>
      </button>
      <button
        class="xl-button xl-button--primary xl-button--size-default"
      >
        <span>
          确定
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Modal 测试visible和onVisibleChange 1`] = `null`;

exports[`Modal 测试visible和onVisibleChange 2`] = `
<div
  class="xl-modal__wrap"
  style="z-index: 1001;"
>
  <div
    class="xl-modal"
    style="width: 520px;"
  >
    <button
      class="xl-modal__icon"
    >
      <span
        class="xl-modal__icon-x"
      >
        <i
          class="xl-icon"
        >
          <svg
            fill="currentColor"
            style="height: 1em; width: 1em;"
            viewBox="0 0 352 512"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
        </i>
      </span>
    </button>
    <div
      class="xl-modal__header"
    >
      <div
        class="xl-modal__title"
      >
        title
      </div>
    </div>
    <div
      class="xl-modal__body"
    >
      content
    </div>
    <div
      class="xl-modal__footer"
    >
      <button
        class="xl-button xl-button--default xl-button--size-default"
      >
        <span>
          取消
        </span>
      </button>
      <button
        class="xl-button xl-button--primary xl-button--size-default"
      >
        <span>
          确定
        </span>
      </button>
    </div>
  </div>
</div>
`;
